### PR TITLE
Add Controller to Devices section so users can add additional pcie-root-ports or virtio-scsi controllers

### PR DESCRIPTION
**Pull request by awels**
Signed-off-by: Alexander Wels <awels@redhat.com>

**What this PR does / why we need it**:
Allow the user to add additional pci or scsi controllers to the VM spec. In order to hot plug additional devices or disks, a pcie-root-port needs to be available. By default only a single extra pcie-root-port is available which is not always sufficient. There is also a limitation on the number of pcie-root-ports one can add to a VM, so for the can one wants to add a large number of disks, the PR allows the user to specify a virtio-scsi controller.

Instead of allowing the user to specify all the little details of the controllers, they can specify 3 fields:
Type: (pci or scsi)
Index: (0 - 31, with 0 being unavailable for pci)
IOThread: The number of iothreads to allocate.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
User can now specify additional pcie-root-ports or virtio-scsi controllers in VM(I) spec
```

### CODE REVIEW

**Comment by rmohr** [comment]
@awels was it considered to avoid that people have to deal with this low level detail? Would it make sense to simply define one or the other way a hotplug maximum and the rest would be done internally?

**Comment by awels** [comment]
I was actually afraid I was abstracting away too much this way. For instance for performance reason on might actually want multiple scsi controllers with a large number of disks, to sort of split the load. Initially I was just going to create a virtio-scsi controller in response to enabling hot plugging. But it was pointed out that depending on the workload people might want different things. Heck people might want to mix and match scsi and virtio hot plugged disks. Which is why I settled on this approach. I am not opposed to what you are suggesting, but that might not give enough control in certain situations.

**Comment by awels** [comment]
@rmohr How about we only give them explicit controller types. One for virtio, and one for virtio-scsi. And if we find a need for more knobs we can add them later.

**Comment by rmohr** [comment]
> @rmohr How about we only give them explicit controller types. One for virtio, and one for virtio-scsi. And if we find a need for more knobs we can add them later.

Exposing controllers at all is a big change. It adds a lot of complexity and a lot of knowledge needs to be present to make use of it properly. It is a difference if I expose devices for different use-cases or if I expose such more vm-architecture-related backend controllers. This may introduce difficulties for users and the project (increased compatibiliy testing matrix for instance).

 I think we need more voices here. @davidvossel @vladikr what do you think?

**Comment by mhenriks** [comment]
I think I would be in favor of a higher level API and not exposing controllers directly.

**Comment by davidvossel** [comment]
> I think we need more voices here. @davidvossel @vladikr what do you think?

I think the general idea is fine, we just need to make the mapping to our api a little more obvious. I think we can simplify it a bit.

From a super high level, what we're wanting to do is create expansion for the number of disks that can be attached (hotplugged) to a running VMI. What's the field here on the disk that's most critical to this expansion? I think it's the disk's `bus` right?

So, In order for this to make sense to the user, why not have a hotplug expansion field on our api that makes what's being configured more obvious. 

take this current example
```yaml
spec:
  domain:
    devices:
      controllers:
      - type: pci
        index: 1
        ioThreads: true
      - type: scsi
        index: 0
```

Maybe we could represent it like
```yaml
spec:
  domain:
    devices:
      diskExpansionSlots:
      - bus: virtio
        slots: 32
      - bus: scsi
        slots: 100
```

and we'd calculate everything behind the scenes on the libvirt domain xml.

I'm not saying the api i outlined there is exactly what we should go with or anything. I'm just trying to get us to think about how to create something that enables hotplug expansion and maps obviously to the values already present in our api around how disks are represented.

**Comment by rmohr** [comment]
> ```yaml
> spec:
>   domain:
>     devices:
>       diskExpansionSlots:
>       - bus: virtio
>         slots: 32
>       - bus: scsi
>         slots: 100
> ```
> I'm not saying the api i outlined there is exactly what we should go with or anything. I'm just trying to get us to think about how to create something that enables hotplug expansion and maps obviously to the values already present in our api around how disks are represented.

This looks already better than exposing the controller directly. Still I think this means a lot to know about VMs just to be able to e.g. hotplug at some point one or two disks to do a mariadb backup. There were tuning-aspects brought in, why the controllers would need to  be exposed. I am not sure if we need something like this at all if a normal user just wants to perform hotplugs.

We could for instance do one of these things:
 * Use a default architecture which allows X hotplugs by default. If we allow right now e.g. 100 hotplugs by default, we probably don't have to expose anything to users with the current pod-spwaning approach.
 * Add a simple api field which allow specifying the number of the maximum hotplugable disks on the api.

**Comment by awels** [comment]
A few things to note, for a scsi bus the number of disks that can be added is effectively unlimited. So for most practical cases you will only need a single virtio-scsi controller. Now if you are expecting a lot of disks, adding more scsi controllers if for balancing the load between them. For a virtio bus the limit on the number of disks that can be added is the max number of pcie-root-ports that are available. And those slots are shared between all devices, so we can't just say well lets just use 10 of them so we handle most cases, because other devices might want to use the slots. As for tuning, the biggest thing I care about is the IOThreads.

How about:
```yaml
spec:
  domain:
    devices:
      diskExpansionSlots:
      - bus: virtio
        slots: 32
      - bus: scsi
        ioThreads: 4
```
With slots being optional, if not specified its 1? And IoThreads being optional as well.

**Comment by rmohr** [comment]
> ```yaml
>         ioThreads: 4
> ```

What is the difference between iothreads on the controller vs iothreads on the disks?

**Comment by awels** [comment]
That is a good question, it translates to iothreads on the driver. I am fine with not exposing it, and if we find a use case that really needs it, adding it later.
```xml
  <controller type='scsi' index='0' model='virtio-scsi'>
    <driver iothread='4'/>
    <address type='pci' domain='0x0000' bus='0x00' slot='0x0b' function='0x0'/>
  </controller>
```

**Comment by davidvossel** [comment]
> That is a good question, it translates to iothreads on the driver. I am fine with not exposing it, and if we find a use case that really needs it, adding it later.

should setting the `ioThreadsPolicy=[auto|shared]` tunable somehow relate to what we set on the driver during the domain xml conversion similarly to how it does to the disk iothreads?

**Comment by davidvossel** [comment]
here's another thought that perhaps we should consider.

Why not simply limit hotplug to scsi for the time being which has a much higher device count.  So for now we'd be saying that if you want hotplug, you have to use scsi. We'd then always create the scsi controller and all vmis would be hot pluggable by default. 

can we get away with that for now?

**Comment by vladikr** [comment]
Sorry, I'm late to review this. 
The `iothread` setting on the controller indicates to which iothread this controller will be assigned.
For example, if we've defined 4 IOThreads for this VMI, we can place this controller on a specific one.
I don't think we should set this right now. I don't see a specific use-case for doing this.

**Comment by vladikr** [comment]
> > That is a good question, it translates to iothreads on the driver. I am fine with not exposing it, and if we find a use case that really needs it, adding it later.
> 
> should setting the `ioThreadsPolicy=[auto|shared]` tunable somehow relate to what we set on the driver during the domain xml conversion similarly to how it does to the disk iothreads?

It shouldn't. Unless we want to assign this controller to a specific IOThread.

**Comment by vladikr** [comment]
From my perspective, KubeVirt should be able to automatically configure the right controllers and set the disks on it.

For virtio-scsi, we can simply let users specify `bus: virtio-scsi` on the disks. We know how many disks are defined on the VMI, so we can simply create one scsi controller with the virt-scsi driver and place all the following disks on this controller (up to 256 per controller)

For pci disks, we know that there is a limit... can't we add additional pcie-root-ports on the fly?

**Comment by awels** [comment]
So according to the documentation it is not possible to add pcie-root-ports on the fly, which why we need to define them ahead of time. We had some discussion on how to proceed from here, and I would like to summarize the discussion:
1. We will always add the vrtio-scsi by default.
2. Provide a mechanism to opt out of this for certain situations like high performance VMs.
3. When hot plugging disks, we need to ensure that we are plugging them into the correct controller which we should be able to do during the domaim xml generation.

I will update this PR to do step 1 and 2. I also need to coordinate that effort with #4408 because we are adding a virtio-scsi controller in that PR.

File: docs/vm-configuration.md [hunk 1]
```diff
@@ -441,3 +441,35 @@ spec:
...
+        iothread: 4
+      - type: scsi
+        index: 0
+```
+In the above example we added 2 additional controllers, one pcie-root-port at index 1 (Index 0 is reserved
+for pcie-root, when type is pci) and one virtio-scsi controller at index 0. KubeVirt will fill in the correct
+model based on the type. pci model is pcie-root-port and scsi will be virtio-scsi. Index is a required field
+and for pci should be between 1 and 31, for scsi between 0 and 31.
+
+iothread is an optional field assigning iothreads to the controller. When used in general the number of iothreads
```
Comment (davidvossel) [inline docs/vm-configuration.md:474]: just a reminder that this documentation isn't accurate anymore.

**Review by davidvossel — COMMENTED** [review]
The new direction of this PR makes sense to me. Code looks good and everything.  I'd suggest moving this logic to your HotPlug PR now though instead of treating this logic independently. I think it made sense to have this discussion independent of HotPlug, but since this is specific to the HotPlug logic, i think it should all go in together.


> I will update this PR to do step 1 and 2. I also need to coordinate that effort with #4408 because we are adding a virtio-scsi controller in that PR.

yup, once #4408  is merged, i think it's as simple as adding an if statement to the "needsSCSIController()" function.

**Comment by awels** [comment]
Closing in favor #4458

**Pull request closed without merging**